designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,955 @@
+1. [Customize merging of config options](https://github.com/eslint/eslint/issues/9192)
+1. [specifying ignore patterns in the config](https://github.com/eslint/eslint/issues/10891) (implemented in `.eslintrc` after this RFC was written)
+1. [specifying `--ext` information in the config](https://github.com/eslint/eslint/issues/11223) (implemented in `.eslintrc` after this RFC was written)
+1. [using `extends` in `overrides`](https://github.com/eslint/eslint/issues/8813) (implemented in `.eslintrc` after this RFC was written)
+
+The only reason that these requests are difficult to implement in ESLint is because of how complex the `.eslintrc` configuration format is. Any changes made to any part of `.eslintrc` processing end up affecting millions of ESLint installations, so we have ended up stuck. The complicated parts of `.eslintrc` include:
+
+1. Resolution behavior of modules (`plugins`, `parser`, `extends`)
+1. Merging of cascading config files in a directory structure
+1. Merging of config files via `extends`
+1. Overriding of configuration via `overrides`
+
+All of these complexities arise from `.eslintrc` format of describing what should happen rather than how it should happen. We keep trying to anticipate how users want things to happen and guessing wrong has led to increased complexity.
+
+This RFC proposes a fresh start for configuration in ESLint that takes into account all of the requests we've received over the years. By simplifying and stripping configuration we can actually make a configuration system that is flexible enough to accomodate changes in the future and easy enough to implement that we won't be afraid to do so.
+
+## Detailed Design
+
+Design Summary:
+
+1. Introduce a new `eslint.config.js` configuration file format
+1. Searching for `eslint.config.js` starts from the current working directory and continues up
+1. All `eslint.config.js` files are treated as if they have `root: true`
+1. There is no automatic merging of config files
+1. The `eslint.config.js` configuration is not serializable and objects such as functions and plugins may be added directly into configuration
+1. An `@eslint/eslintrc` utility is provided to aid with backwards compatibility
+1. Remove the concept of environments (`env`) (including `--env`, `/*eslint-env*/`, and `env` in config files)
+1. Remove `.eslintignore` and `--ignore-file` (no longer necessary)
+1. Remove `--rulesdir`
+1. Remove `--ext`
+1. Remove `--resolve-plugins-relative-to`
+1. Replace `context.parserPath` with `context.parser` (the path to the parser is no longer valid)
+1. Allow plugins to specify a `parsers` key so parsers can be referenced the same as rules and processors in plugins.
+
+### The `eslint.config.js` File
+
+The `eslint.config.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports an object: 
+
+```js
+module.exports = {
+    name: "name",
+    files: ["*.js"],
+    ignores: ["*.test.js"],
+    settings: {},
```

Comment:
As I've been thinking about this more, the property name `settings` feels out of place. Is there any way we can make give it a less vague name? Maybe `pluginOptions`?

Reply:
I'm not sure if changing the name has much of a benefit vs. the confusion it could potentially cause, but I don't feel very strongly about it.

Reply:
So after looking into this some more, I think changing `settings` to something else will make backwards compatibility more challenging (these are exposed through `context.settings` in rules). While I agree the name isn't ideal, I think it would cause more trouble to change it than just to let it be.

Reply:
That's fair :+1: